node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,450 @@
+				}
+
+				if let Some(peer) = peers.first() {
+					net.write_notification(peer.clone(), message).await;
+				}
+			}
+			Action::ReportPeer(peer, rep) => {
+				net.report_peer(peer, rep).await
+			}
+			Action::StartWork(relay_parent) => {
+				live_heads.push(relay_parent);
+				if let Some(view_update)
+					= update_view(&peers, &live_heads, &net, &mut local_view).await
+				{
+					if let Err(_) = dispatch_update_to_all(
+						view_update,
+						event_producers.values(),
+						&mut ctx,
+					).await {
+						log::warn!("Aborting - Failure to dispatch messages to overseer");
+						return
+					}
+				}
+			}
+			Action::StopWork(relay_parent) => {
+				live_heads.retain(|h| h != &relay_parent);
+				if let Some(view_update)
+					= update_view(&peers, &live_heads, &net, &mut local_view).await
+				{
+					if let Err(_) = dispatch_update_to_all(
+						view_update,
+						event_producers.values(),
+						&mut ctx,
+					).await {
+						log::warn!("Aborting - Failure to dispatch messages to overseer");
+						return
+					}
+				}
+			}
+
+			Action::PeerConnected(peer, role) => {
+				match peers.entry(peer.clone()) {
+					Entry::Occupied(_) => continue,
+					Entry::Vacant(vacant) => {
+						vacant.insert(PeerData {
+							view: View(Vec::new()),
+						});
+
+						if let Err(_) = dispatch_update_to_all(
+							NetworkBridgeEvent::PeerConnected(peer, role),
+							event_producers.values(),
+							&mut ctx,
+						).await {
+							log::warn!("Aborting - Failure to dispatch messages to overseer");
+							return
+						}
+					}
+				}
+			}
+			Action::PeerDisconnected(peer) => {
+				if peers.remove(&peer).is_some() {
+					if let Err(_) = dispatch_update_to_all(
+						NetworkBridgeEvent::PeerDisconnected(peer),
+						event_producers.values(),
+						&mut ctx,
+					).await {
+						log::warn!("Aborting - Failure to dispatch messages to overseer");
+						return
+					}
+				}
+			},
+			Action::PeerMessages(peer, messages) => {
+				let peer_data = match peers.get_mut(&peer) {
+					None => continue,
+					Some(d) => d,
+				};
+
+				let mut outgoing_messages = Vec::with_capacity(messages.len());
+				for message in messages {
+					match message {
+						WireMessage::ViewUpdate(new_view) => {
+							if new_view.0.len() > MAX_VIEW_HEADS {
+								net.report_peer(peer.clone(), MALFORMED_VIEW_COST).await;
+								continue
+							}
+
+							if new_view == peer_data.view { continue }
+							peer_data.view = new_view;
+
+							let update = NetworkBridgeEvent::PeerViewChange(
+								peer.clone(),
+								peer_data.view.clone(),
+							);
+
+							outgoing_messages.extend(
+								event_producers.values().map(|producer| producer(update.clone()))
+							);
+						}
+						WireMessage::ProtocolMessage(protocol, message) => {
+							let message = match event_producers.get(&protocol) {
+								Some(producer) => Some(producer(
+									NetworkBridgeEvent::PeerMessage(peer.clone(), message)
```

Comment:
No, the `PeerMessage` is only broadcast to the subsystem that has registered the `ProtocolId` that the message was under.

Reply:
The `message.clone()` I am more worried about is in `write_notification`, where we will be sending large messages to each peer. There are probably a few more copies down the line deeper in the network code, but it would be nice to send  a `Bytes` in `write_notification` instead of `Vec<u8>`